scylla/src/history.rs

```diff
@@ -432,3 +432,649 @@ fn write_fiber_attempts(fiber: &FiberHistory, f: &mut std::fmt::Formatter<'_>) -
+
+        let mut query = Query::new("SELECT * FROM system.local");
+        let history_collector = Arc::new(HistoryCollector::new());
+        query.set_history_listener(history_collector.clone());
+
+        session.query(query.clone(), ()).await.unwrap();
+
+        let history: StructuredHistory = history_collector.clone_structured_history();
+
+        let displayed = "Queries History:
+=== Query #0 ===
+| start_time: 2022-02-22 20:22:22 UTC
+| Non-speculative attempts:
+| - Attempt #0 sent to 127.0.0.1:19042
+|   request send time: 2022-02-22 20:22:22 UTC
+|   Success at 2022-02-22 20:22:22 UTC
+|
+| Query succesful at 2022-02-22 20:22:22 UTC
+=================
+";
+        assert_eq!(
+            displayed,
+            format!(
+                "{}",
+                set_one_db_error_message(set_one_node(set_one_time(history)))
+            )
+        );
+
+        // Prepared queries retain the history listener set in Query.
+        let prepared = session.prepare(query).await.unwrap();
+        session.execute(&prepared, ()).await.unwrap();
+
+        let history2: StructuredHistory = history_collector.clone_structured_history();
+
+        let displayed2 = "Queries History:
+=== Query #0 ===
+| start_time: 2022-02-22 20:22:22 UTC
+| Non-speculative attempts:
+| - Attempt #0 sent to 127.0.0.1:19042
+|   request send time: 2022-02-22 20:22:22 UTC
+|   Success at 2022-02-22 20:22:22 UTC
+|
+| Query succesful at 2022-02-22 20:22:22 UTC
+=================
+=== Query #1 ===
+| start_time: 2022-02-22 20:22:22 UTC
+| Non-speculative attempts:
+| - Attempt #0 sent to 127.0.0.1:19042
+|   request send time: 2022-02-22 20:22:22 UTC
+|   Success at 2022-02-22 20:22:22 UTC
+|
+| Query succesful at 2022-02-22 20:22:22 UTC
+=================
+";
+        assert_eq!(
+            displayed2,
+            format!(
+                "{}",
+                set_one_db_error_message(set_one_node(set_one_time(history2)))
+            )
+        );
+    }
+
+    #[tokio::test]
+    async fn failed_query_history() {
+        let uri = std::env::var("SCYLLA_URI").unwrap_or_else(|_| "127.0.0.1:9042".to_string());
+        let session = SessionBuilder::new().known_node(uri).build().await.unwrap();
+
+        let mut query = Query::new("This isnt even CQL");
+        let history_collector = Arc::new(HistoryCollector::new());
+        query.set_history_listener(history_collector.clone());
+
+        assert!(session.query(query.clone(), ()).await.is_err());
+
+        let history: StructuredHistory = history_collector.clone_structured_history();
+
+        let displayed =
+"Queries History:
+=== Query #0 ===
+| start_time: 2022-02-22 20:22:22 UTC
+| Non-speculative attempts:
+| - Attempt #0 sent to 127.0.0.1:19042
+|   request send time: 2022-02-22 20:22:22 UTC
+|   Error at 2022-02-22 20:22:22 UTC
+|   Error: Database returned an error: The submitted query has a syntax error, Error message: Error message from database
```

Comment:
As @piodul once mentioned, relying on error messages in tests can be pain when they change, but it seems inevitable here.

Reply:
I try to mitigate this by setting the error message to a constant string using `set_one_db_error_message`. The test will break only once our internal error printing breaks, which will be a code change anyway.